### ISSUE TITLE
Increase eligibility of Instrinsic Element names in JSX/TSX

### ISSUE THREAD

**Issue by ryansolid** [microsoft/TypeScript#31606]
## Search Terms
JSX, compilation, Intrinsic Elements

## Suggestion
Allow more flexibility in defining Intrinsic Elements in JSX. The lowercase/uppercase convention works for React but doesn't represent the range of the spec. Any name should be allowed to be registered as an Intrinsic Element.

## Use Cases
JSX is just an XML like syntax language that gets compiled down to JavaScript. How that compilation works and what it outputs could be different depending on the library. Assuming case or character set to determine Intrinsic Elements outside the spec makes it much more difficult to support TypeScript for other JSX libraries.

## Examples
 Suppose a library wanted wanted to allow control flow to be precompiled using a $ symbol to prevent interference with the range of Value Elements the end user could use:
```jsx
<$ each={ items }>{ item => <div>{ item.title }</div>}</$>
```
This is perfectly valid JSX and could be compiled away before runtime.

Or how about something like NativeScript which uses a PascalCase syntax for it's intrinsic elements? It should be perfectly fine to set aside a few of these without having mess with per file imports or messing with Global scope.

I mean given the trend to compile down, and idea very much at the heart of JSX, it seems incredibly limiting to restrict this to a convention, no matter how popular it is. If the spec allows so should Typescript.

Yet today, you cannot set $ nor Button as an Intrinsic JSX elements.

## Checklist

My suggestion meets these guidelines:

* [ ] ??? This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

**Comment by nmain** [comment]
Deciding whether `<Foo />` is emitted as `h(Foo)` or `h("Foo")` using `JSX.IntrinsicElements` would be type-directed emit.  But there are JSX transforms out there with non-trivial differences from React, so maybe the better answer would be richer compiler options for `jsx` and `jsxFactory`?

Alternatively, you can set `jsx: "preserve"` and let Babel sort it out.

**Comment by ryansolid** [comment]
Yeah I could see how that could be useful. But to be clear I wrote this issue primarily for just Type Checking reasons.  Even if you use `preserve` to handle the compilation yourself which I think is reasonable here, you have to try to hack around to try to get Type Checking to work.

I think given the emit behavior perhaps it makes sense why Type Checking has been so restrictive historically. But now with interopt with tools like Babel 7 and preset-typescript, I'm perfectly content for more complicated (non-HyperScript generating) scenarios, to use `preserve`. I think on a purest level jsxFactory would be really powerful, but I imagine that will take a little bit more doing. I'd love just to see the Type Checking fixed to start as it's the blocker. Having TypeScript not know how to emit custom JS is more than forgivable.

**Comment by trusktr** [comment]
> Alternatively, you can set jsx: "preserve" and let Babel sort it out.

That's already what I do with @ryansolid's [babel-preset-solid](https://www.npmjs.com/package/babel-preset-solid) (TypeScript preserves JSX, then Babel does another pass).

But the main point is about how how TypeScript works with JSX types. It would be great if not only did it support custom intrinsics like `<$ />`, but that it also allowed the type of a JSX expression to be based on the root element used in the expressions.

So for example the inferred type of `el` in the following example would come out to `HTMLDivElement` (given some way for us to define this before hand):

```ts
const el = <div>Hello!</div> // el is of type HTMLDivElement
```

And for example:

```ts
const el = <p><span>Hello!</span></p> // el is of type HTMLParagraphElement
```

It would be great to specify what types of children an element can have too. For example, `<select>` should only accept `<option>` elements. It'd be great if...

```ts
const el = <select><div></div></select> // ... this were a type error...
const el2 = <select><option ...>...</option></select> // ... and this were OK.
```

This would be interesting for libraries that compile JSX input to DOM elements for example (instead outputting createElement factory calls or similar).

To make the idea more clear, the compiled output of

```js
const el = <div>Hello!</div>
console.log(el instanceof HTMLDivElement) // true
```

could be something like

```js
const el = function() {
  const el = document.createElement('div')
  div.textContent = 'Hello!'
  return div
}()
console.log(el instanceof HTMLDivElement) // true
```
